WalletWasabi.Fluent/ViewModels/Dialog/DialogViewModelBaseOfTResult.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.Reactive.Linq;
+using System.Threading.Tasks;
+using ReactiveUI;
+
+namespace WalletWasabi.Fluent.ViewModels.Dialog
+{
+    /// <summary>
+    /// Base ViewModel class for Dialogs that returns a value back.
+    /// </summary>
+    /// <typeparam name="TResult">The type of the value to be returned when the dialog is finished.</typeparam>
+    public abstract class DialogViewModelBase<TResult> : DialogViewModelBase
+    {
+        protected DialogViewModelBase()
+        {
+            this.WhenAnyValue(x => x.IsDialogOpen)
+                .DistinctUntilChanged()
+                .Subscribe(OnIsDialogOpenChanged);
+        }
+
+        private void OnIsDialogOpenChanged(bool obj)
+        {
+			// Trigger when closed abruptly (via the Overlay or the back button).
+            if (!obj & CurrentTaskCompletionSource is { } & !DialogReturnedWithValue)
+            {
+				 Close();
+            }
+        }
+
+        private TaskCompletionSource<TResult> CurrentTaskCompletionSource { get; set; }
+
+        /// <summary>
+        /// Method to be called when the dialog intends to close
+        /// without returning a value.
+        /// </summary>
+        protected void Close()
+        {
+            if (CurrentTaskCompletionSource is null)
+            {
+                throw new InvalidOperationException("CloseDialog with value return method failed due to missing TCS.");
+            }
+
+            CurrentTaskCompletionSource.SetResult(default);
+            CurrentTaskCompletionSource = null;
+            IsDialogOpen = false;
+
+            OnDialogClosed();
+        }
+
+        /// <summary>
+        /// Method to be called when the dialog intends to close
+        /// and ready to pass a value back to the caller.
+        /// </summary>
+        /// <param name="value">The return value of the dialog</param>
+        protected void Close(TResult value)
```

Comment:
lets only have a Close TResult method... people can use Unit if they dont want it.

Reply:
or Close (TResult value = default)

we have duplicated code otherwise.